### PR TITLE
Overloading

### PR DESCRIPTION
**Pull request by tavisca-ashwini**
@neilunadkat : Kindly review the code.

@ytokas : Kindly review the code.

### CODE REVIEW

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+using OperatorOverloading.Model;
+
+namespace OperatorOverloading.Host
+{
+    class Operation
+    {
+       static void Main(string[] args)
+        {
+           double money = 0.0;
+           string tempCurrency = " ";
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: default currency value should be null or string.Empty

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+           Money money1 = new Money();
+           Money money2 = new Money();
+           Money money3 = new Money();
+
+           Console.WriteLine("Enter Amount for Money1");
+           money = Convert.ToDouble(Console.ReadLine());
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: use double.tryparse

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+           Console.WriteLine("Enter Currency for Money1");
+           tempCurrency = Convert.ToString(Console.ReadLine());
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: no need to convert string to string

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+           Console.WriteLine("Enter Amount for Money2");
+           money = Convert.ToDouble(Console.ReadLine());
+           Console.WriteLine("Enter Currency for Money2");
+           tempCurrency = Convert.ToString(Console.ReadLine());
+           
+           money2.Amount = money;
+           money2.Currency = tempCurrency;
+
+           if (money1.Amount < 0 || money2.Amount < 0)
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: amount can be negative

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+           {
+               Console.WriteLine("Enter valid Amount");
+               return;
+           }
+           try
+           {
+               money3 = money1 + money2;
+           }
+           catch (InvalidCurrencyException ex) // catched exception
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: handle general exceptions

File: 01 Operator Overloading/OperatorOverloading.Model/InvalidCurrencyException.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OperatorOverloading.Model
+{
+    public class InvalidCurrencyException : Exception
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/InvalidCurrencyException.cs]: make this serializable
use Message property of Exception

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloading.Model
+{
+    public class Money
+    {
+        private double amount;
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: private var should start with '_'

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        private string currency;
+
+        public double Amount
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: public double Amount {get; set;}

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            get
+            {
+                return this.currency;
+            }
+            set
+            {
+                this.currency = value;
+            } 
+        }
+        public Money(int money, string tempCurrency)
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: var name should tell what it is

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            this.amount = money;
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: don't set private variables

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        public Money()
+        {
+
+        }
+        public static Money operator +(Money Money1, Money Money2)
+        {
+            Money Money3 = new Money();
+
+            if (Money1.Currency.Equals(Money2.Currency))
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: check for case sensitivity

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            {
```
Comment (tavisca-ashwini) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: check for overflow

**Comment by tavisca-ashwini** [comment]
@ytokas  : Please review the updated code.

File: 01 Operator Overloading/OperatorOverloading.Host/Operation.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+           Money money3 = new Money();
+
+           /*obtaining values or first object*/
+           Console.WriteLine("Enter Amount for first object");
+           while (!double.TryParse(Console.ReadLine(), out money))
+           {
+               Console.WriteLine(" Enter valid entry for money");
+           }
+           Console.WriteLine("Enter Currency for first object");
+           tempCurrency = Console.ReadLine();
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Host/Operation.cs]: Add null and empty checks for currency. Also check for string length, it should be 3 chars.

File: 01 Operator Overloading/OperatorOverloading.Model/InvalidCurrencyException.cs [hunk 14]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OperatorOverloading.Model
+{
+    [Serializable]
+    public class InvalidCurrencyException : Exception
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/InvalidCurrencyException.cs]: just adding the Serializable attribute will not make it serializable. You need to implement the proper constructor.

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            this.Amount = money;
+            this.Currency = tempCurrency;
+        }
+        public Money()
+        {
+
+        }
+
+        //operator + is overloaded
+        public static Money operator +(Money Money1, Money Money2)
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: parameters names should be camel casing

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            string currency;
+            double money = 0.0;
+            Money Money3 = new Money();
+
+            if (!string.IsNullOrEmpty(Money1.Currency) && !string.IsNullOrEmpty(Money2.Currency) && string.Equals(Money1.Currency, Money2.Currency, StringComparison.OrdinalIgnoreCase))
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: What will happen if money1 or money2 is null

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                currency = Money1.Currency.ToUpper();
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: Where is this used?

File: 01 Operator Overloading/OperatorOverloading.Model/Money.cs [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+                Money3.Amount = Money1.Amount + Money2.Amount;
+                Money3.Currency = Money1.Currency;
+
+                if (Money1.Amount <= (double.MaxValue - Money2.Amount))
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/Money.cs]: Check double.IsInfinity()

File: 01 Operator Overloading/OperatorOverloading.Model/OverflowException.cs [hunk 19]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloading.Model
+{
+    [Serializable]
+    class OverflowException : Exception
```
Comment (ytokas) [inline 01 Operator Overloading/OperatorOverloading.Model/OverflowException.cs]: make it serializable

File: 01 OperatorOverloading/OperatorOverloading.Host/Operation.cs [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+           tempCurrency = Console.ReadLine();
+           
+           money2.Amount = money;
+           money2.Currency = tempCurrency;
+
+           try
+           {
+               money3 = money1 + money2;
+           }
+           catch (InvalidCurrencyException ex) // catched exception
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Operation.cs]: If you are going to print the Exception message every time, catch only System.Exception.

File: 01 OperatorOverloading/OperatorOverloading.Host/Operation.cs [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+               return;
+           }
+           catch (OverflowException e)
+           {
+               Console.WriteLine("" + e.Message);
+               Console.WriteLine("");
+               Console.ReadKey();
+               return;
+           }
+           Console.WriteLine("Amount entered :  " +money1.Amount + " Currency for Amount :" +money1.Currency);
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Operation.cs]: Try overriding ToString method of Money to print the details.

File: 01 OperatorOverloading/OperatorOverloading.Model/InvalidCurrencyException.cs [hunk 22]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OperatorOverloading.Model
+{
+    [Serializable]
+    public class InvalidCurrencyException : Exception
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/InvalidCurrencyException.cs]: This is not serializable

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        //operator + is overloaded
+        public static Money operator +(Money Money1, Money Money2)
+        {
+            string currency;
+            double money = 0.0;
+            Money Money3 = new Money();
+
+            if (!string.IsNullOrEmpty(Money1.Currency) && !string.IsNullOrEmpty(Money2.Currency) && string.Equals(Money1.Currency, Money2.Currency, StringComparison.OrdinalIgnoreCase))
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: Null check for money objects.

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                currency = Money1.Currency.ToUpper();
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: what is this for?

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+                Money3.Amount = Money1.Amount + Money2.Amount;
+                Money3.Currency = Money1.Currency;
+
+                if (Money1.Amount <= (double.MaxValue - Money2.Amount))
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: see double.IsInfinity() to check overflow

File: 01 OperatorOverloading/OperatorOverloading.Model/OverflowException.cs [hunk 26]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloading.Model
+{
+    [Serializable]
+    class OverflowException : Exception
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/OverflowException.cs]: make it serializable

**Comment by tavisca-ashwini** [comment]
@ytokas  : Please check updated code.
(serealizable is yet to apply)

**Comment by ytokas** [comment]
Please make all the changes asked in the review before submitting the pull.

**Comment by tavisca-ashwini** [comment]
@ytokas : Please check updated code.

**Comment by ytokas** [comment]
I dont want to see review request until you are done with all the changes. If you have any questions come and ask me.

**Comment by tavisca-ashwini** [comment]
@ytokas  : Please check the code for review

**Comment by tavisca-ashwini** [comment]
@ytokas : please review the updated code

**Comment by ytokas** [comment]
Move the code to OperationOverload folder and raise pull request.

**Comment by tavisca-ashwini** [comment]
@ytokas  : Moved the code in Operator|Overloading folder. kindly review it.

**Comment by tavisca-ashwini** [comment]
@ytokas : Moved the code in OperatorOverloading folder. Kindly review it.

**Pull request merged**